### PR TITLE
Reduce number of surfaces required when presenting platform views

### PR DESCRIPTION
**Pull request by knopp**
Fixes https://github.com/flutter/flutter/issues/129710
Fixes https://github.com/flutter/flutter/issues/138936

Implements https://flutter.dev/go/optimized-platform-view-layers

## Example

This scene would normally require 5 surfaces, but with the PR it comes down to 2 (when drawing over platform views) and 1 when all drawing is outside of platform views).

https://github.com/flutter/flutter/assets/96958/091da832-bfbc-44a2-8da5-d55d84024c96

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @cbracken @knopp Are we going to make progress on this?

**Comment by knopp** [comment]
https://github.com/flutter/engine/pull/42960 just got merged today. I'll rebase this tomorrow, retest and remove the wip.

**Comment by knopp** [comment]
I'm not sure why `verifyb143464703_soft_noxform` fails. I tried running it locally but I get all golden failures, even through images seemingly look same. Maybe slightly difference in antialiasing?

**Comment by chinmaygarde** [comment]
The failure should point to both the expectation and the actual image. Can you paste both of them here. If its AA, we should be able to reason about it.

**Comment by chinmaygarde** [comment]
Were you able to make progress on figuring out the test failure?

**Comment by knopp** [comment]
Sorry, been busy with other stuff. Adding WIP until I get back to this.

**Comment by zanderso** [comment]
@knopp friendly ping. Still planning to work on this?

**Comment by knopp** [comment]
So, here is the failure: 
```
Expected:/home/matej/Projects/flutter/engine/src/out/host_debug_unopt/gen/flutter/shell/platform/embedder/assets/expectation_verifyb143464703_soft_noxform.png
Got:/home/matej/Projects/flutter/engine/src/out/host_debug_unopt/gen/flutter/shell/platform/embedder/assets/actual_verifyb143464703_soft_noxform.png
```
I ran it on x86 linux machine because on arm64 all golden tests fail :-/ Here are the images:

Expectation:
expectation_verifyb143464703_soft_noxform

Actual:
actual_verifyb143464703_soft_noxform

I thnk there might be some very slight change of antialiasing? These golden test seem really fragile TBH. Maybe it would be worth snapping the widget on physical pixel boundaries to remove the antialiasing, which would hopefully make this work across architectures.

**Comment by knopp** [comment]
This reduces the number of surfaces in [this example](https://github.com/flutter/flutter/issues/138936) from 21 (depending on window size) to 1 or 2 (if overlaid by scroll bar or header), fixing the issue. Each of the platform view in example has a Flutter border, which currently results in an intermediate layer. However because those borders are always painted below the platform views, they can all be collapsed into the base layer.

File: shell/platform/embedder/embedder_render_target_cache.cc [hunk 1]
```diff
@@ -10,63 +10,41 @@ EmbedderRenderTargetCache::EmbedderRenderTargetCache() = default;
...
-      compatible_targets.pop();
-      resolved_render_targets[view.first] = std::move(target);
-    }
+std::unique_ptr<EmbedderRenderTarget>
+EmbedderRenderTargetCache::GetRenderTarget(
+    const EmbedderExternalView::RenderTargetDescriptor& descriptor) {
+  auto compatible_target = cached_render_targets_.find(descriptor);
+  if (compatible_target == cached_render_targets_.end()) {
+    return nullptr;
+  } else {
```
Comment (cbracken) [inline shell/platform/embedder/embedder_render_target_cache.cc:19]: pedantic style nit: just leave the `if` as a bail-out conditional, then make the contents of the `else` the main function body. Avoids a teensy bit of indentation, but also slightly defensive against any future change that would leave the top-level without a `return` statement.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 2]
```diff
@@ -122,13 +121,309 @@ static FlutterBackingStoreConfig MakeBackingStoreConfig(
...
+    assert(view->HasPlatformView());
+    view_identifier = view->GetViewIdentifier();
+    params = view->GetEmbeddedViewParams();
+
+    clipped_frame = view->GetEmbeddedViewParams()->finalBoundingRect();
+    SkMatrix transform;
+    for (auto i = params->mutatorsStack().Begin();
+         i != params->mutatorsStack().End(); ++i) {
+      const auto& m = *i;
+      if (m->GetType() == MutatorType::kTransform) {
```
Comment (cbracken) [inline shell/platform/embedder/embedder_external_view_embedder.cc:143]: Use a switch statement here; it's wordier, but helps with listing.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 3]
```diff
@@ -122,13 +121,309 @@ static FlutterBackingStoreConfig MakeBackingStoreConfig(
...
+
+  std::vector<SkIRect> coverage() {
+    return flutter_contents_region_.getRects();
+  }
+
+ private:
+  std::vector<PlatformView> platform_views_;
+  std::vector<EmbedderExternalView*> flutter_contents_;
+  DlRegion flutter_contents_region_;
+  friend class LayerBuilder;
```
Comment (cbracken) [inline shell/platform/embedder/embedder_external_view_embedder.cc:257]: nit: Move this to the bottom.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 4]
```diff
@@ -122,13 +121,309 @@ static FlutterBackingStoreConfig MakeBackingStoreConfig(
...
+      target = render_target_cache_.GetRenderTarget(
+          EmbedderExternalView::RenderTargetDescriptor(
+              SkISize{static_cast<int32_t>(config.size.width),
+                      static_cast<int32_t>(config.size.height)}));
+    }
+    if (target != nullptr) {
+      return target;
+    } else {
+      return create_render_target_callback_(context, aiks_context, config);
+    }
```
Comment (cbracken) [inline shell/platform/embedder/embedder_external_view_embedder.cc:425]: Similar comment to before; we can just leave the if as a bailout and unwrap the else to the default path for the function body after the if.

**Comment by knopp** [comment]
> this PR remove the ability to use Opacity(..) around an AppKitView() also it seems that each time a new platform view is created the UI blink during one or 2 frames.

I can't reproduce that even with your example. This is `AppKitView` wrapped with 0.1 opacity:

Screenshot 2023-12-06 at 09 47 00

That said, once opacity is enabled, I can definitely reproduce the flicker. I have reopened https://github.com/flutter/flutter/issues/138936.

**Pull request merged**